sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/ArrayExpressionUtils.java

```diff
@@ -116,58 +116,131 @@ public class ArrayExpressionUtils {
     return SQLOrderingUtil.compareDoubles(c1, c2);
   };
 
+  // boolean
+  // foldable optimize
+  public static int binarySearchNullSafe(Boolean[] data, Boolean value) {
```

Comment:
I think it's better to take `ArrayData` here to simplify the expression implementation. We can call `arrayData.toBooleanArray` or `.toObjectArray` for non-nullable and nullable arrays.

Reply:
For nullable arrays, I don't think concrete types can help the performance. `Object[]` should be OK.

Reply:
Okay